Swiften/Examples/MIXListAndJoin/MIXListAndJoin.cpp

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright (c) 2017 Tarun Gupta
+ * Licensed under the simplified BSD license.
+ * See Documentation/Licenses/BSD-simplified.txt for more information.
+ */
+
+#include <iostream>
+#include <memory>
+#include <unordered_set>
+
+#include <boost/optional.hpp>
+
+#include <Swiften/Elements/DiscoInfo.h>
+#include <Swiften/Elements/ErrorPayload.h>
+#include <Swiften/Elements/MIXParticipant.h>
+#include <Swiften/Client/Client.h>
+#include <Swiften/Client/ClientOptions.h>
+#include <Swiften/Client/ClientXMLTracer.h>
+#include <Swiften/Disco/GetDiscoItemsRequest.h>
+#include <Swiften/EventLoop/SimpleEventLoop.h>
+#include <Swiften/JID/JID.h>
+#include <Swiften/MIX/MIX.h>
+#include <Swiften/MIX/MIXImpl.h>
+#include <Swiften/MIX/MIXRegistry.h>
+#include <Swiften/Network/BoostNetworkFactories.h>
+#include <Swiften/Network/Timer.h>
+#include <Swiften/Network/TimerFactory.h>
+#include <Swiften/Queries/GenericRequest.h>
+
+#include <Swiften/Base/Log.h>
+
+using namespace Swift;
+using namespace std;
+
+static SimpleEventLoop eventLoop;
+static BoostNetworkFactories networkFactories(&eventLoop);
+
+static std::shared_ptr<Client> client;
+static MIXImpl::ref mix;
+static JID mixChannelJID;
+static JID mixServiceDomain;
+static std::unordered_set<std::string> supportedNodes;
+static MIXRegistry* mixRegistry_;
+
+static void handleChannelJoined(const JID& jid) {
+    // Successfully joined the channel.
+    assert(mixRegistry_);
+    assert(mixRegistry_->getMIXInstance(jid));
+
+    std::cout << "Successfully joined channel " << jid << std::endl;
+
+    //trying to rejoin the same channel joined. Should not send the iq request. Will print a warning.
+    mixRegistry_->joinChannel(mixChannelJID, supportedNodes);
+
+    mix = mixRegistry_->getMIXInstance(jid);
+
+    // Can perform other functions with MIXImpl object: mix.
+
+    // Finally leave channel.
+    // This is commented to check the sync between different users using different clients(resources) to connect.
+    //mixRegistry_->leaveChannel(mixChannelJID);
+}
+
+static void handleChannelJoinFailed(ErrorPayload::ref /*error*/) {
+    std::cout << "Channel Join Failed." << std::endl;
+}
+
+static void handleChannelLeft(const JID& jid) {
+    assert(mixRegistry_);
+    assert(!mixRegistry_->getMIXInstance(jid));
+
+    std::cout << "Successfully left channel " << jid << std::endl;
+}
+
+static void handleRosterPopulated() {
+    XMPPRoster* xmppRoster = client->getRoster();
+    int channelCount = 0;
+    std::cout << "Already Joined channels: " << std::endl;
+    for (auto item : xmppRoster->getItems()) {
+        channelCount++;
```

Comment:
Seems like a sensible idea. I wanted to hold my client till onInitialRosterPopulated is called and syncRegistryWithRoster is successfully executed.